### DISCUSSION TITLE
Revamp public facing API

### DISCUSSION THREAD

**Discussion by Kralizek** [AutoFixture/AutoFixture#1188] [Ideas]
This issue comes from the conversation we had in #1175.

The idea is to make the public API self-discoverable and easier for new users while retaining the same functionalities available to power users.

The conversation started as creating simple extension methods on the `IFixture` interface but, if we are willing to introduce breaking changes, we can be more daring.

My idea is to align to modern APIs (especially those belonging to the Microsoft extensions libraries) where there is a builder and a service.

In code terms, I'd split `IFixture` into these two interfaces (simplified)

```csharp
public interface IFixtureBuilder
{
    IFixture Build();

    void Add(ICustomization customization);
}

public interface IFixture
{
    object Create(Type type);
}
```

This would make users have:

```csharp
var builder = new FixtureBuilder();
builder.Add(new MyDomainCustomization());
builder.AddAutoMoq(); // extension method
builder.Customize<Something>(o => o...); // extension method

var fixture = builder.Build();
var something = fixture.Create<Something>(); // extension method
```

Since each extension method of the IFixtureBuilder interface would return the builder itself, methods can be chained.

```csharp
var builder = new FixtureBuilder()
                                 .AddAutoMoq()
                                 .Customize<Something>(o => o...);

var fixture = builder.Build();
```

**Comment by Kralizek** [comment]
Any comment here? @aivascu @zvirja

**Comment by aivascu** [comment]
@Kralizek I like the idea to be able to customize the `Fixture` using extension methods, provided by third-party libraries (internal or third-party).
My personal preference would be something more asp.net-like in terms of API.
```csharp
IFixture AddMoq(this IFixture fixture, Action<MoqConfiguration> configure);
IFixture AddMoq(this IFixture fixture, MoqConfiguration configure);
IFixture AddMoq(this IFixture fixture)
```
This would allow for more options, depending on the client preference.

**Comment by Kralizek** [comment]
@aivascu i take you don't fancy the idea of the builder like other asp.net extension libraries do ( configuration, dependency injection and logging above all)

**Comment by aivascu** [comment]
What would be the benefit of implementing a fixture builder?

I think the builder pattern could apply here, if by introducing it we could optimize the process of "building" the `Fixture` instance. 
In asp.net the container builder has the purpose to do exactly that. It builds the dependency tree and then it becomes immutable in order to ensure the reliability and performance of instance resolution.

Looking at the current implementation of `Fixture.Customize(ICustomization)`, I see very little room for optimization, since currently it delegates all the work back to the customization and the `Fixture` remains mutable after it is built.

I think we can borrow only the parts that would benefit us most. In this particular case it's more about the syntactic sugar rather than having an actual ioc container.

**Comment by Kralizek** [comment]
@aivascu 

The purpose of introducing the builder is to split the interfaces so that users don't get confused on when to use what.

Registrations and configurations would extend `IFixtureBuilder`, consumption would be done through `IFixture`.

The idea came mostly from the feedback in the original issues where @zvirja didn't want to overcharge `IFixture` with too many methods or extensions. So, rather than introducing `Extend()`, I thought the builder was a cleaner pattern.

**Comment by aivascu** [comment]
@Kralizek I see your point, but I don't think users would be enticed to use either the builder or anything `.Extend()` would expose. Why would anyone use an additional abstraction, when you have direct access to `IFixture`?
I fear that to make AutoFixture more intellisense friendly and not overburden the `IFixture` interface we'd have to redesign the type system.

**Comment by Kralizek** [comment]
@aivascu 

I thought that the reason why #1175 was closed was to be able to have a more ambitious plan in `v5` when making intellisense friendly the surface API.

Now, I see there are three approaches to make AutoFixture friendlier:
1. We split `IFixture` in logical parts, the builder idea is just one of many ways
2. We introduce a facade like the `Extend()` method suggested by @zvirja (I personally agree with you and dislike this option)
3. We introduce extension methods on `IFixture` as it is today, basically #1175.

Honestly, it feels like we're stuck in the land between "we don't want to release anything half-baked" and "we don't want to introduce too many changes" like we are afraid to gets our hands dirty and start baking.

**Comment by zvirja** [comment]
@aivascu @Kralizek If we want to "rethink our extensibility API" and introduce [yet another standard](https://xkcd.com/927/), we should think of properties of this API. Among others, one of the requirements I see is API discoverability. As of today, `IFixture` and `Fixture` implement a lot of other interfaces and completion is very cluttered. If we just add yet another extension method to `IFixture`, I don't believe that it's ultimately more discoverable than `fixture.Customize(new AutoFixture.Moq.Auto....)`. After we migrate other extensions to the same pattern and add even more customizations, `fixture.` completion scope will be really cluttered.

That was my motivation why I suggested a brand new extension point `fixture.Extend()`. This way the completion items are solely related to the fixture v2 customization API and nothing else. The suggested builder patter provides the same benefits, but it introduces another entity which is more complicated to naturally incorporate.

**Comment by Kralizek** [comment]
@zvirja 

> If we just add yet another extension method to `IFixture`, I don't believe that it's ultimately more discoverable than `fixture.Customize(new AutoFixture.Moq.Auto....)`.

I still believe that `fixture.AddAutoMoq()` is much easier to discover than `fixture.Customize(new AutoFixture.Moq.Auto....)`.

Especially if you are a new user who doesn't know where to look at what.
